src/main/java/com/google/firebase/database/connection/NettyWebSocketClient.java

```diff
@@ -0,0 +1,202 @@
+              p.addLast(sslContext.newHandler(ch.alloc(), uri.getHost(), DEFAULT_WSS_PORT));
+              p.addLast(
+                  new HttpClientCodec(),
+                  new HttpObjectAggregator(8192),
+                  WebSocketClientCompressionHandler.INSTANCE,
+                  channelHandler);
+            }
+          });
+
+      ChannelFuture channelFuture = bootstrap.connect(uri.getHost(), DEFAULT_WSS_PORT);
+      this.channel = channelFuture.channel();
+      channelFuture.addListener(
+          new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture future) throws Exception {
+              if (!future.isSuccess()) {
+                eventHandler.onError(future.cause());
+              }
+            }
+          }
+      );
+    } catch (SSLException e) {
+      eventHandler.onError(e);
+    }
+  }
+
+  @Override
+  public void close() {
+    checkState(channel != null, "channel not initialized");
+    try {
+      channel.close();
+    } finally {
+      group.shutdownGracefully();
+      executorService.shutdown();
+      // TODO(hkj): https://github.com/netty/netty/issues/7310
+    }
+  }
+
+  @Override
+  public void send(String msg) {
+    checkState(channel != null && channel.isActive(), "channel not connected for sending");
+    channel.writeAndFlush(new TextWebSocketFrame(msg));
+  }
+
+  /**
+   * Handles low-level IO events. These events fire on the firebase-websocket-worker thread. We
+   * notify the {@link WebsocketConnection} on all events, which then hands them off to the
+   * RunLoop for further processing.
+   */
+  private static class WebSocketClientHandler extends SimpleChannelInboundHandler<Object> {
+
+    private final WebsocketConnection.WSClientEventHandler delegate;
+    private final WebSocketClientHandshaker handshaker;
+
+    WebSocketClientHandler(
+        URI uri, String userAgent, WebsocketConnection.WSClientEventHandler delegate) {
+      this.delegate = checkNotNull(delegate, "delegate must not be null");
+      checkArgument(!Strings.isNullOrEmpty(userAgent), "user agent must not be null or empty");
+      this.handshaker = WebSocketClientHandshakerFactory.newHandshaker(
+          uri, WebSocketVersion.V13, null, true,
+          new DefaultHttpHeaders().add("User-Agent", userAgent));
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext context) {
+      // Do nothing
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext context) {
+      handshaker.handshake(context.channel());
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext context) {
+      delegate.onClose();
+    }
+
+    @Override
+    public void channelRead0(ChannelHandlerContext context, Object message) throws Exception {
+      Channel channel = context.channel();
+      if (!handshaker.isHandshakeComplete()) {
+        checkState(message instanceof FullHttpResponse);
```

Comment:
This is running on the firebase-websocket-worker thread, right? What happens when we throw an exception? Does it get handled, logged, etc?

Reply:
Tested by injecting some exceptions. It fires the `NettyWebSocketClient.exceptionCaught()` event, which we hand off to `delegate.onError()`. The net result is, the existing connection gets closed gracefully, and the client moves on to the next connection attempt.

Reply:
If this check fails, that means there's a  bug in this code and we should at minimum log it (and crashing the process wouldn't be out-of-the-question).  So if this just silently retries without logging anything, then I don't think that's sufficient.

Reply:
`delegate.onError()` logs the exception (delegate is `WebsocketConnection.WSClientHandlerImpl` in this case):

```
      if (e.getCause() != null && e.getCause() instanceof EOFException) {
        logger.error("WebSocket reached EOF", e);
      } else {
        logger.error("WebSocket error", e);
      }
```

Reply:
👍